File: Loenn/entities/RalseiDummyAttacherTrigger.lua
local ralseiDummyAttacherTrigger = {}

ralseiDummyAttacherTrigger.name = "MaggyHelper/RalseiDummyAttacherTrigger"
ralseiDummyAttacherTrigger.placements = {
    name = "trigger",
    data = {
        hoverOffsetX = 20,
        hoverOffsetY = -10,
        hoverSpeed = 2.0,
        hoverAmplitude = 4.0,
        once = true,
        removeOnExit = false
    }
}

ralseiDummyAttacherTrigger.fieldInformation = {
    hoverOffsetX = {
        fieldType = "number",
        minimumValue = -100,
        maximumValue = 100
    },
    hoverOffsetY = {
        fieldType = "number",
        minimumValue = -100,
        maximumValue = 100
    },
    hoverSpeed = {
        fieldType = "number",
        minimumValue = 0.1,
        maximumValue = 10.0
    },
    hoverAmplitude = {
        fieldType = "number",
        minimumValue = 0.0,
        maximumValue = 20.0
    },
    once = {
        fieldType = "boolean"
    },
    removeOnExit = {
        fieldType = "boolean"
    }
}

return ralseiDummyAttacherTrigger

File: Code/Triggers/RalseiDummyAttacherTrigger.cs
using Celeste;
using Celeste.Entities;
using Celeste.Mod;
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;

namespace Celeste.Triggers
{
    /// <summary>
    /// Trigger that attaches a floating Ralsei dummy to the player.
    /// Place this trigger in a level to automatically attach the Ralsei dummy when the player enters.
    /// </summary>
    [CustomEntity("MaggyHelper/RalseiDummyAttacherTrigger")]
    [Tracked]
    public class RalseiDummyAttacherTrigger : Trigger
    {
        private float hoverOffsetX = 20f;
        private float hoverOffsetY = -10f;
        private float hoverSpeed = 2f;
        private float hoverAmplitude = 4f;
        private bool once = true;
        private bool removeOnExit = false;
        private bool triggered = false;

        public RalseiDummyAttacherTrigger(EntityData data, Vector2 offset) : base(data, offset)
        {
            hoverOffsetX = data.Float("hoverOffsetX", 20f);
            hoverOffsetY = data.Float("hoverOffsetY", -10f);
            hoverSpeed = data.Float("hoverSpeed", 2f);
            hoverAmplitude = data.Float("hoverAmplitude", 4f);
            once = data.Bool("once", true);
            removeOnExit = data.Bool("removeOnExit", false);
        }

        public override void OnEnter(Player player)
        {
            if (once && triggered)
                return;

            triggered = true;

            // Check if player already has the attacher
            var existingAttacher = player.Get<RalseiDummyAttacher>();
            if (existingAttacher == null)
            {
                // Create and add the attacher
                var attacher = new RalseiDummyAttacher();
                attacher.SetHoverOffset(hoverOffsetX, hoverOffsetY);
                attacher.SetHoverSpeed(hoverSpeed);
                attacher.SetHoverAmplitude(hoverAmplitude);
                player.Add(attacher);
            }
            else
            {
                // Update existing attacher settings
                existingAttacher.SetHoverOffset(hoverOffsetX, hoverOffsetY);
                existingAttacher.SetHoverSpeed(hoverSpeed);
                existingAttacher.SetHoverAmplitude(hoverAmplitude);
            }

            if (once)
            {
                RemoveSelf();
            }
        }

        public override void OnLeave(Player player)
        {
            if (removeOnExit)
            {
                player.RemoveRalseiDummy();
            }
        }
    }
}